vmware-event-router/internal/provider/vcsim/vcsim.go

```diff
@@ -0,0 +1,187 @@
+		return nil, fmt.Errorf("invalid %s credentials: username and password must be set", config.BasicAuth)
+	}
+
+	username := cfg.Auth.BasicAuth.Username
+	password := cfg.Auth.BasicAuth.Password
+	parsedURL.User = url.UserPassword(username, password)
+
+	client, err := govmomi.NewClient(ctx, parsedURL, cfg.InsecureSSL)
+	if err != nil {
+		return nil, errors.Wrap(err, "create vCenter simulator client")
+	}
+
+	l := log.New(os.Stdout, color.Fata("[VCSIM] "), log.LstdFlags)
+	vcsim.Logger = l
+	vcsim.client = *client
+
+	// apply options (overwrite any defaults)
+	for _, opt := range opts {
+		opt(&vcsim)
+	}
+
+	// seed the metrics stats
+	vcsim.stats = metrics.EventStats{
+		Provider:    string(config.ProviderVCSIM),
+		Type:        config.EventProvider,
+		Address:     cfg.Address,
+		Started:     time.Now().UTC(),
+		EventsTotal: new(int),
+		EventsErr:   new(int),
+		EventsSec:   new(float64),
+	}
+
+	go vcsim.PushMetrics(ctx, ms)
+	return &vcsim, nil
+}
+
+// Stream implements the event provider interface and starts the event stream
+func (vcsim *EventStream) Stream(ctx context.Context, processor processor.Processor) error {
+	mgr := event.NewManager(vcsim.client.Client)
+	defer func() {
+		// ignore error against vcsim
+		_, _ = mgr.Destroy(ctx)
+	}()
+
+	const (
+		pageSize = 10
+		tail     = true
+		force    = true
+	)
+
+	// get events for all objects
+	ref := vcsim.client.ServiceContent.RootFolder
+	handler := eventHandler(ctx, vcsim, processor)
+
+	// blocks
+	return mgr.Events(ctx, []types.ManagedObjectReference{ref}, pageSize, tail, force, handler)
+}
+
+func eventHandler(_ context.Context, vcsim *EventStream, proc processor.Processor) func(moRef types.ManagedObjectReference, baseEvents []types.BaseEvent) error {
+	var (
+		errCount int
+		source   = vcsim.client.URL().String()
+	)
+
+	return func(_ types.ManagedObjectReference, baseEvents []types.BaseEvent) error {
+		if len(baseEvents) == 0 {
+			return nil
+		}
+
+		// reverse slice because vcsim sends events in descending key order
+		reverse(baseEvents)
+		for _, e := range baseEvents {
+			ce, err := events.NewCloudEvent(e, source)
+			if err != nil {
+				vcsim.Printf("skipping event %v because it could not be converted to CloudEvent format: %v", e, err)
+				errCount++
+				continue
+			}
+
+			err = proc.Process(*ce)
+			if err != nil {
+				vcsim.Printf("could not process event %v: %v", ce, err)
+				errCount++
+				continue
+			}
+		}
+
+		// update metrics
+		vcsim.Lock()
+		total := *vcsim.stats.EventsTotal + len(baseEvents)
+		vcsim.stats.EventsTotal = &total
+		errTotal := *vcsim.stats.EventsErr + errCount
+		vcsim.stats.EventsErr = &errTotal
+		vcsim.Unlock()
+
+		return nil
+	}
+}
+
+// reverse reverses the given slide
+func reverse(events []types.BaseEvent) {
+	for i := len(events)/2 - 1; i >= 0; i-- {
+		opp := len(events) - 1 - i
+		events[i], events[opp] = events[opp], events[i]
+	}
+}
+
+// Shutdown closes the underlying connection to vCenter simulator
```

Comment:
Is this function supposed to log out of govmomi client?

Reply:
Yes, but also see the comment `// EventManager:EventManager does not implement: Destroy_Task`

Reply:
Ok, that makes sense now that I realize you're trying to implement an interface.